server/src/core/invitation/invitation-service.ts

```diff
@@ -0,0 +1,131 @@
+
+const COLLECTION = 'invitations';
+
+export class Invitations implements InvitationService {
+  private db: Db;
+  private collection: Collection<DbInvitation>;
+  private indexesCreated: boolean;
+
+  constructor(db: Db) {
+    this.db = db;
+    this.collection = this.db.collection(COLLECTION);
+    this.indexesCreated = false;
+  }
+
+  async createIndexes(): Promise<void> {
+    if (this.indexesCreated) return;
+
+    try {
+      /*
+      Expire every invitation document 
+      after 172800 seconds (i.e., two days) have passed since creation
+      */
+      await this.collection.createIndex(
+        { expiresAt: 1},
+        { expireAfterSeconds: 172800 }
+      );
+      
+      this.indexesCreated = true;
+    }
+    catch (e) {
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async create(args: InvitationCreateArgs): Promise<Invitation> {
+    const { invitorId, invitorName, inviteeName, inviteePhone, inviteeEmail, inviteeRole, hasAccount } = args;
+    const now = new Date();
+    const invitation: DbInvitation = {
+      _id: generateId(),
+      invitorId, 
+      invitorName, 
+      inviteeName,
+      inviteePhone,
+      inviteeEmail,
+      inviteeRole,
+      hasAccount,
+      status: 'pending',
+      expiresAt: now,
+      createdAt: now,
+      updatedAt: now,
+    }
+
+    try {
+      const res = await this.collection.insertOne(invitation);
+      return res.ops[0];
+    }
+    catch (e) {
+      rethrowIfAppError(e);
+
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async accept(invitationId: string): Promise<Invitation> {
+    try {
+      const result = await this.collection.findOneAndUpdate(
+        { _id: invitationId },
+        { $set: { status: 'accepted' } },
+        { upsert: true, returnOriginal: false }
+      );
+      
+      if (!result) throw createResourceNotFoundError(messages.ERROR_INVITATION_NOT_FOUND);
+      
+      return result.value;
+    }
+    catch(e) {
+      rethrowIfAppError(e);
+
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async reject(invitationId: string): Promise<Invitation> {
+    try {
+      const result = await this.collection.findOneAndUpdate(
+        { _id: invitationId },
+        { $set: { status: 'rejected' } },
+        { upsert: true, returnOriginal: false }
+      );
+
+      if (!result) throw createResourceNotFoundError(messages.ERROR_INVITATION_NOT_FOUND);
+
+      return result.value;
+    }
+    catch(e) {
+      rethrowIfAppError(e);
+
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async getAllByUser(userId: string, userPhone: string): Promise<Invitation[]> {
+    validators.validatesGetAllByUser({ userId, userPhone });
+    try {
+      const result = await this.collection.find({ $or: [{ invitorId: userId }, { inviteePhone: userPhone }] }).sort({ createdAt: -1 }).toArray();
+      return result;
+    }
+    catch (e) {
+      throw createDbOpFailedError(e.message);
+    }
+  }
```

Comment:
I see a potential bug here. Since the user can fetch all the invitations he/she has sent out, he/she also has access to the invitation ids. This means that he/she is able to accept/reject invitations that he/she has sent. This definitely not something that should be possible. Maybe it would have been a good idea to have a separate secret code after all, and to hide that code from the invitation sender.

But a much quicker fix is to return the list of invitations without including the id. i.e. include a projection in this db query that excludes the _id from the returned results (`project({ _id: 0 })`).